cli: add timeouts to `node status` and `node ls`

tbg opened the pull request:
GRPC in its recent versions has become quite noisy and in particular does
not play well with context cancellation.

Mute it in all client commands except when actually starting a server.

Refactor the cli package to avoid the previously common mixing of stoppers
and contexts. Nobody really needs the stopper usually (and if so, they
could make their own); it is simpler to just have a context and cancel it
when done.

Touches #16489 (by making it more straightforward to use timeouts).

Fixes #20308.

tbg committed:
cli: remove spurious GRPC warnings

GRPC in its recent versions has become quite noisy and in particular does not play well
with context cancellation.

Mute it in all client commands except when actually starting a server.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20308)
<!-- Reviewable:end -->

tbg committed:
cli: don't mix contexts and stoppers

Refactor the cli package to avoid the previously common mixing of stoppers and
contexts. Nobody really needs the stopper usually (and if so, they could make
their own); it is simpler to just have a context and cancel it when done.

Touches #16489 (by making it more straightforward to use timeouts).

tbg committed:
lint: skip TestHelpURLs in lintshort

tbg committed:
cli: add --timeout to `node ls|status`

Defaults to no timeout.

Both of these commands internally do a KV scan, which
may never finish on a sufficiently broken cluster.

Release note (cli change): the `node status` and `node ls` commands now support a timeout.

Updates #16489.

nvanbenschoten commented:
Very nice cleanup getting rid of the stopper mess!

---

Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 7 of 7 files at r3, 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/cli/error.go, line 67 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20308#-L-5bhoM_jNx0uMiW8CV:-L-5bhoM_jNx0uMiW8CW:b-3zfyme) ([raw file](https://github.com/cockroachdb/cockroach/blob/52a29f6d3fcdc967c153452fd60de224089fe38a/pkg/cli/error.go#L67)):*
> ```Go
> 		case *net.OpError:
> 			return connDropped()
> 		default:
> ```

nit: no need for the default cases here and below. I'd remove them from all three switches and move the `return err` out.

---

*[pkg/cli/init.go, line 60 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20308#-L-5ZK1dH0li5XGr82eX:-L-5ZK1dH0li5XGr82eY:bew0xq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/768b4d83a87f3035aa21b1d87e0700b087ddb126/pkg/cli/init.go#L60)):*
> ```Go
> }
> 
> func getInitClient() (serverpb.InitClient, *stop.Stopper, error) {
> ```

why delete this one but leave `getStatusClient`?

---

*[pkg/cli/start.go, line 949 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20308#-L-5_MIWsS9Cubi9X-W7:-L-5_MIWsS9Cubi9X-W8:b-czvjk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5260069034cd33f837e3e53b82dcda66152fb99c/pkg/cli/start.go#L949)):*
> ```Go
> 	}
> 	stopper.AddCloser(stop.CloserFn(func() {
> 		_ = conn.Close()
> ```

How was this handled before? Did we just never close the connection?

---

*[pkg/cli/start.go, line 969 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20308#-L-5_3oaDKgceYiiVYVq:-L-5_3oaDKgceYiiVYVr:bmlbhcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/5260069034cd33f837e3e53b82dcda66152fb99c/pkg/cli/start.go#L969)):*
> ```Go
> }
> 
> func stopperContext(stopper *stop.Stopper) context.Context {
> ```

Do we still need this function?

---

*[pkg/util/grpcutil/log.go, line 33 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20308#-L-5XRDTXKViBTM6lPA7:-L-5XRDTXKViBTM6lPA8:b-6levtf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5980a4f86b485d007baf1baf98b38f450cbb02a1/pkg/util/grpcutil/log.go#L33)):*
> ```Go
> )
> 
> var discardLogger = grpclog.NewLoggerV2(ioutil.Discard, ioutil.Discard, ioutil.Discard)
> ```

I'd expect that we would at least want to print errors to stderr outside of starting a cluster. When you say that gRPC has become noisy, do you mean that it's printing spurious errors or warnings?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20308#-:-L-5c2aCNY2F5rrfZw3D:b-xonka3)*
<!-- Sent from Reviewable.io -->

tbg committed:
cli: remove stopperContext

tschottdorf commented:
Review status: 1 of 17 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/cli/error.go, line 67 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20308#-L-5bhoM_jNx0uMiW8CV:-L-8KhcjQ7wFCItpkiZ7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/52a29f6d3fcdc967c153452fd60de224089fe38a/pkg/cli/error.go#L67)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: no need for the default cases here and below. I'd remove them from all three switches and move the `return err` out.
</blockquote></details>

Done.

---

*[pkg/cli/init.go, line 60 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20308#-L-5ZK1dH0li5XGr82eX:-L-8L-AkEpnvLF5zhg42:b-5s5gtm) ([raw file](https://github.com/cockroachdb/cockroach/blob/768b4d83a87f3035aa21b1d87e0700b087ddb126/pkg/cli/init.go#L60)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

why delete this one but leave `getStatusClient`?
</blockquote></details>

It had two callers, but just inlined it there. Removed.

---

*[pkg/cli/start.go, line 949 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20308#-L-5_MIWsS9Cubi9X-W7:-L-8L39acP8ibEZswr9R:b-wd0qtq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5260069034cd33f837e3e53b82dcda66152fb99c/pkg/cli/start.go#L949)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

How was this handled before? Did we just never close the connection?
</blockquote></details>

That's what it seems like to me.

---

*[pkg/cli/start.go, line 969 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20308#-L-5_3oaDKgceYiiVYVq:-L-8MjL_xQy8ZszbVLFJ:bhvq9pr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5260069034cd33f837e3e53b82dcda66152fb99c/pkg/cli/start.go#L969)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we still need this function?
</blockquote></details>

Removed the remaining callers and put this method to sleep.

---

*[pkg/util/grpcutil/log.go, line 33 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20308#-L-5XRDTXKViBTM6lPA7:-L-8bVgv5yuGO5vqPmxg:b2b1pzt) ([raw file](https://github.com/cockroachdb/cockroach/blob/5980a4f86b485d007baf1baf98b38f450cbb02a1/pkg/util/grpcutil/log.go#L33)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'd expect that we would at least want to print errors to stderr outside of starting a cluster. When you say that gRPC has become noisy, do you mean that it's printing spurious errors or warnings?
</blockquote></details>

Yeah, it prints a bunch of warnings. Some of this is because using context cancellation tears down everything eagerly, but I don't think it's just that. How do you like the latest iteration?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20308)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r6, 6 of 7 files at r7, 6 of 6 files at r8, 2 of 2 files at r9, 5 of 5 files at r10.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/util/grpcutil/log.go, line 57 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/20308#-L-8gYrrB41L1fBoyMWP:-L-8gYrseVB0BLb-94-g:b9dc9m8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bdd3d1ef82bd1962c46eecb4b8a48240e1c84f3/pkg/util/grpcutil/log.go#L57)):*
> ```Go
> type logger log.Severity
> 
> func (severity *logger) Info(args ...interface{}) {
> ```

Do these receivers need to be pointers?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20308#-:-L-8gwlYpS4irzHafHQg:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
cli: grpc logs at WARNING for start, ERROR for rest

This avoids muddling up client output with noisy GRPC connection chatter,
but lease some amount of logging in place for hard errors (which I have
never seen in practice).

tbg committed:
cli: wait for connection cleanup

I refactored this to rely solely on context cancellation in
a previous commit. Unfortunately, we run cli tools multiple
times in the same process during tests, and we reset globals
that are picked up by the RPCContext used by open connections,
which results in data races.

Return a closure from getClientGRPCConn which waits for the
Stopper to stop to circumvent this.

tschottdorf commented:
Had to unfortunately reintroduce waiting for the stopper, see latest commit.

---

Review status: 1 of 18 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/util/grpcutil/log.go, line 57 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/20308#-L-8gYrrB41L1fBoyMWP:-L-A-Qf6jrIPPSqj3AhZ:b-oyaayu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bdd3d1ef82bd1962c46eecb4b8a48240e1c84f3/pkg/util/grpcutil/log.go#L57)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do these receivers need to be pointers?
</blockquote></details>

Yeah, see the comment above (synthetic stack frames).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20308#-:-L-A-UsqM2GQESJAmPAy:ba38of8)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR!
